apps/manager_app/device_authentication.c

```diff
@@ -0,0 +1,343 @@
+      uint8_t *challenge =
+          (uint8_t *)&(query->auth_device.challenge.challenge[0]);
+      result->auth_device = sign_random_challenge(challenge);
+      next_state = RESULT;
+      break;
+    }
+    case MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG: {
+      /* If SIGN_RANDOM_NUM == state, but the request_type received is
+       * MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG, it's an unexpected step in
+       * the flow. The device will treat it as an attempt to force device
+       * authentication status */
+      device_auth_handle_response(false);
+      result->auth_device = send_flow_complete();
+      next_state = FLOW_COMPLETE;
+      break;
+    }
+    default: {
+      usb_clear_event();
+      break;
+    }
+  }
+
+  return next_state;
+}
+
+static device_auth_state_e result_handler(const manager_query_t *query,
+                                          manager_result_t *result) {
+  pb_size_t request_type = get_request_type(&query->auth_device);
+  device_auth_state_e next_state = RESULT;
+
+  switch (request_type) {
+    case MANAGER_AUTH_DEVICE_REQUEST_INITIATE_TAG:
+    case MANAGER_AUTH_DEVICE_REQUEST_CHALLENGE_TAG: {
+      usb_clear_event();
+      break;
+    }
+    case MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG: {
+      bool verified = query->auth_device.result.verified;
+      device_auth_handle_response(verified);
+      result->auth_device = send_flow_complete();
+      next_state = FLOW_COMPLETE;
+      break;
+    }
+    default: {
+      usb_clear_event();
+      break;
+    }
+  }
+
+  return next_state;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void device_authentication_flow(const manager_query_t *query) {
+  /* First state of the device authentication would be SIGN_SERIAL_NUMBER */
+  device_auth_state_e state = SIGN_SERIAL_NUM;
+  evt_status_t event = {0};
+
+  /* Query to be decoded on USB event reception */
+  manager_query_t decoded_query = MANAGER_QUERY_INIT_ZERO;
+
+  bool valid_query = true;
+
+  while (FLOW_COMPLETE != state) {
+    manager_result_t result =
+        get_manager_result_template(MANAGER_RESULT_AUTH_DEVICE_TAG);
+
+    /* When this loop runs for the first time, this check will pass as qeury is
+     * already checked before user confirmation. */
+    if (false == valid_query) {
+      event = get_events(EVENT_CONFIG_USB, MAX_INACTIVITY_TIMEOUT);
+      P0_HANDLER(event);
+
+      decode_manager_query(
+          event.usb_event.p_msg, event.usb_event.msg_size, &decoded_query);
+      query = &decoded_query;
+    }
+
+    if (false == check_manager_request(query, MANAGER_QUERY_AUTH_DEVICE_TAG)) {
+      valid_query = false;
+      continue;
+    } else {
+      valid_query = true;
+    }
+
+    switch (state) {
+      case SIGN_SERIAL_NUM: {
+        state = sign_serial_handler(query, &result);
+        break;
+      }
```

Comment:
reject and exit the flow if invalid query is received in this state.

Reply:
fixed